GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -117,9 +124,81 @@ internal void FillTreeViewNode(TreeNode treeViewNode)
             public Node? LastNode => _nodesList.Count > 0 ? _nodesList[_nodesList.Count - 1] : null;
         }
 
-        private abstract class Tree : IDisposable
+        /// <summary>A common base class for both <see cref="Node"/> and <see cref="Tree"/>.</summary>
+        internal abstract class NodeBase
+        {
+            /// <summary>The child nodes.</summary>
+            protected internal Nodes Nodes { get; protected set; }
+
+            /// <summary>The corresponding tree node.</summary>
+            protected internal virtual TreeNode TreeViewNode { get; set; }
+
+            /// <summary>
+            /// Marks this node to be included in multi-selection. See <see cref="Select(bool, bool)"/>.
+            /// This is remembered here instead of relying on the status of <see cref="TreeViewNode"/>
+            /// because <see cref="Nodes.FillTreeViewNode(TreeNode)"/> recycles <see cref="TreeNode"/>s
+            /// and may change the association between <see cref="Node"/> and <see cref="TreeNode"/>.
+            /// </summary>
+            protected internal bool IsSelected { get; set; }
+
+            protected internal void Select(bool select, bool includingDescendants = false)
+            {
+                IsSelected = select;
+                ApplyStyle(); // toggle multi-selected node style
+
+                // recursively process descendants if required
+                if (includingDescendants && this.HasChildren())
+                {
+                    foreach (var child in Nodes)
+                    {
+                        child.Select(select, includingDescendants);
+                    }
+                }
+            }
+
+            #region style / appearance
```

Comment:
We generally avoid regions.

Reply:
I see. I used them to group members by context in an attempt to improve the top-level readability of the types. Should I just drop the regions then? Or use other methods like a partial class definitions to achieve the desired grouping?

Reply:
_Personally_ I tend to follow the old .NET Framework coding guidelines https://docs.microsoft.com/en-us/archive/blogs/brada/internal-coding-guidelines. But with hundreds of contributors code styling in this project varies.... We just generally don't use regions :)

Reply:
@RussKie I understand and will keep this in mind. Since you didn't answer my question whether I should drop the regions I've added or extract them to partial class definitions, I'll leave them in place for now.

Reply:
Sounds good.